src/StreamJsonRpc/Reflection/MessageFormatterProgressTracker.cs

```diff
@@ -0,0 +1,168 @@
+        /// </summary>
+        /// <param name="value">The object which should implement <see cref="IProgress{T}"/> to create the <see cref="ProgressParamInformation"/> instance.</param>
+        /// <returns>The assigned <see cref="long"/> progres ID.</returns>
+        public long AddProgressObjectToMap(object value)
+        {
+            Requires.NotNull(value, nameof(value));
+
+            if (this.RequestIdBeingSerialized == null)
+            {
+                throw new NotSupportedException("IProgress<T> objects should not be part of any response or notification.");
+            }
+
+            lock (this.ProgressLock)
+            {
+                long progressId = this.NextProgressId++;
+                this.RequestProgressMap.Add(this.RequestIdBeingSerialized.Value, progressId);
+
+                this.ProgressMap.Add(progressId, new ProgressParamInformation(value));
+
+                return progressId;
+            }
+        }
+
+        /// <summary>
+        /// Class used to keep relevant information of an object that implements <see cref="IProgress{T}"/>.
+        /// </summary>
+        public class ProgressParamInformation
+        {
+            /// <summary>
+            /// Initializes a new instance of the <see cref="ProgressParamInformation"/> class.
+            /// </summary>
+            /// <param name="progressObject">The object implementing <see cref="IProgress{T}"/>.</param>
+            public ProgressParamInformation(object progressObject)
+            {
+                Requires.NotNull(progressObject, nameof(progressObject));
+
+                Type iProgressOfTType = MessageFormatterProgressTracker.FindIProgressOfT(progressObject.GetType());
+
+                Verify.Operation(iProgressOfTType != null, Resources.FindIProgressOfTError);
+
+                this.ValueType = iProgressOfTType.GenericTypeArguments[0];
+                this.ReportMethod = iProgressOfTType.GetRuntimeMethod(nameof(IProgress<int>.Report), new Type[] { this.ValueType });
+                this.ProgressObject = progressObject;
+            }
+
+            /// <summary>
+            /// Gets the actual <see cref="Type"/> reported by <see cref="IProgress{T}"/>.
+            /// </summary>
+            public Type ValueType { get; }
+
+            /// <summary>
+            /// Gets the <see cref="MethodInfo"/> of <see cref="IProgress{T}.Report(T)"/>.
+            /// </summary>
+            public MethodInfo ReportMethod { get; }
+
+            /// <summary>
+            /// Gets the instance of the object implementing <see cref="IProgress{T}"/>.
+            /// </summary>
+            public object ProgressObject { get; }
+        }
+
+        /// <summary>
+        /// Class that implements <see cref="IProgress{T}"/> and sends <see cref="ProgressRequestSpecialMethod"/> notification when reporting.
+        /// </summary>
+        public class JsonProgress<T> : IProgress<T>
```

Comment:
>public [](start = 8, length = 6)

Can we make this class private, and add two helper methods to the helper class itself:
```cs
IProgress<T> CreateProgress<T>(object token)
object CreateProgress(Type valueType);
```

The MessagePack formatter can use the generic method while the JSON formatter would use the non-generic overload. The 1-2 lines in the callers that create this (at least in the JSON formatter) are rather complicated, so offering a method to do it for the formatters within the helper class is goodness.

Note the lack of a `JsonRpc` parameter. Since this class shares in the interest of that JsonRpc instance remaining constant, can we take that as a ctor parameter or property on the helper class?